README.md

```diff
@@ -2,7 +2,157 @@
 
 # Prelude
 
+ASON is a data oriented algorithm designed for compact and speedy storage of AssemblyScript objects in a binary format.
 
+There are many Serialization methods out there that can turn a conceptual object into a buffer of some kind, like JSON and protobuf. Each one of these methods has their benefits, and none of them can take advantage of runtime type information provided by the AssemblyScript compiler and the runtime type information.
+
+Another problem with serialization methods is that they tend to be "tree-like" instead of data oriented. To make matters worse, there are no ways for developers to determine field definitions on a given generic reference object at runtime. This is where the ASON algorithm comes in handy.
+
+Instead of storing a tree structure, ASON stores a set of instructions to assemble an object from scratch using the smallest amount of space possible.
+
+# How To Use
+
+Install from npm:
+
+```
+npm install @ason/assembly
+```
+
+Modify your asconfig to include the transform:
+
+```ts
+{
+  ...
+  "transform": ["@ason/transform"]
+}
+```
+
+Import the library and serialize away!
+
+```ts
+import { ASON } from "@ason/assembly";
+
+// serialize can determine type information
+let buffer = ASON.serialize([29, 6, 4] as StaticArray<u8>);
+
+// deserialize must have the type passed (to perform type assertions)
+let result = ASON.deserialize<StaticArray<u8>>(buffer);
+
+assert(result.length == 3);
+assert(result[0] == <u8>29);
+assert(result[0] == <u8>6);
+assert(result[0] == <u8>4);
+```
+
+It's also possible to save heap allocations and re-use a `Serializer` and `Deserializer` object when serializing multiple objects of the same type.
+
+```ts
+import { Serializer, Deserializer } from "@ason/assembly";
+class Vec3 {
+  constructor(
+    public x: f32,
+    public y: f32,
+    public z: f32,
+  ) {}
+}
+
+let result = new Array<StaticArray<u8>>(); // an array of buffers
+let serializer = new ASON.Serializer<Vec3>();
+
+for (let i = 0; i < 10; i++) {
+  result.push(serializer.serialize(new Vec3(1, 2, 3)));
+}
+
+let deserializer = new ASON.Deserializer<Vec3>();
+
+for (let i = 0; i < 10; i++) {
+  let vec = deserializer.deserialize(result[i]);
+  assert(vec); // make sure the reference isn't null
+  assert(vec.x == 1); // check the properties
+  assert(vec.y == 2);
+  assert(vec.x == 3);
+}
+```
+
+# Implementation
+
+In order to assemble an AssemblyScript object, we need a few key pieces of data about our reference.
```

Comment:
The above example has demonstrated serializing an AS object `Vec3` without the need to add this extra information. Is this only required for tree-like structures? (i.e. where there are references)